Remove most/all requirements for withSocketsDo on Windows

ndmitchell opened the pull request:
This patch tries to eliminate the need to call `withSocketsDo` on Windows. While I can't guarantee that it's never required, it certainly fixes some cases (e.g. [the one I hit most recently](https://github.com/snoyberg/http-client/issues/94#issuecomment-70209111) with http-client by @snoyberg), and I believe probably fixes most if-not-all cases. There are three patches:

* Simplify `withSocketsDo`, so there is no afterwards action. The afterwards action called `atexit` anyway, so was pretty pointless to delay. (In reality, the `atexit` action is probably totally useless, but I'll leave it for now, since it isn't harmful either.)
* Make `withSocketsDo` cheaper by using a Haskell-side CAF to avoid repeat initialization. Now an unnecessary `withSocketsDo` on Linux is free, and on Windows is a few instructions, so they can be placed everywhere.
* Add `withSocketsDo` at all the reasonable entry points. The logic I used is to add them just before creating `MkSocket`, and to those functions which call the C library but don't require a socket, e.g. `getAddrInfo`.

I note that Sigbjorn Finne in 2003 tried to eliminate the need to call `withSocketsDo` using two tricks:

* The `trySysCall` function retries with `withSocketsDo` if the answer is a null pointer.
* If some functions return an error code of `WSANOTINITIALISED` then they are retried.

I now believe all the Sigbjorn's calls are redundant - but they're very cheap and can only help, so I've left them.

CC @fegu, who was brainstorming ideas with me.

tibbe commented:
Looks good to me. Could you please make sure you've tested it on a few Windows versions?

> I now believe all the Sigbjorn's calls are redundant - but they're very cheap and can only help, so I've left them.

I say we should kill them. They make the code harder to follow.

ndmitchell commented:
I have tested on Windows 7, and will test on Windows 8 tonight. I will add some patches removing the existing fixes (which I agree are pretty ugly, due to the retry approach).

ndmitchell commented:
I removed all the retry-based uses of withSocketsDo. The ones in `trySysCall` were all totally redundant. The one in `connect` was almost certainly redundant, unless someone outside the package created a `Socket` directly using the constructor. To guarantee there is no regression I added a preemptive `withSocketsDo` before `connect`.

I have not yet tested on Windows 8, but will do so tonight or later in the week.

ndmitchell commented:
Tested on Windows 8.1, works great. That's all the Windows OS's I have access to. I don't suspect there is anything Windows version specific in here through, we're not actually changing the timings of any winsock calls.
